Render as dots, chloropleths, or isochrones

abyrd opened the pull request:
This PR adds bicubic interpolation, and allows rendering the resulting interpolated surface with dots, with a continuous interpolated color scale, or as a hard cutoff (for isochrones).

abyrd committed:
bicubic interpolation, optional pixel color modules

this allows gridualizer to be used for dots, chloropleths, and isochrones

mattwigway commented on lib/dot.js:
```diff
@@ -113,4 +98,176 @@ export default class Dot extends TileLayer.Canvas {
+  
+}
+
+class InterpolatingColorScale {
```
These color scales should all be in separate files.

abyrd replied:
I'll pull all the color scales and interpolators out into separate files.

abyrd replied:
Isn't the one-file-per-class thing a Java convention? In this case it seems neater to just put all the interpolators into one file (module) and all the color maps into another.

mattwigway replied:
I'll defer to @trevorgerhardt on this but our general convention has been one class per file. In any case they shouldn't be in the same file as the leaflet layer.

abyrd replied:
I checked with @trevorgerhardt and he concurred that one class per file is Java-style and we shouldn't hesitate to export multiple classes per module.

trevorgerhardt replied:
@abyrd my mistake! I meant to refer only to the possibilities of `import`/`export` in ES6. If you are declaring a class, then I would still recommend one class per file.

abyrd replied:
All classes have been split into separate files.

mattwigway commented on lib/dot.js:
```diff
@@ -113,4 +98,176 @@ export default class Dot extends TileLayer.Canvas {
+
+class InterpolatingColorScale {
+
+  // Pass in an array of [[z, r, g, b, a], [z, r, g, b, a], [z, r, g, b, a]...]
```
To match code elsewhere, can we have an array of breaks and a parallel array of colors that can be in any format accepted by `d3-color`?

abyrd replied:
Sure, but I just need some more context for how this is supposed to be used. Can you point me to some other places where the new classes will need to be compatible? Do we really need to support several color formats or can we just use one throughout? Or do you mean that d3-color can be used to normalize color breaks supplied in many formats into RGBA?

mattwigway replied:
Yes, that's correct. `d3-color` parses many color formats and normalizes them. It's not so much a compatibility thing as a consistency with other code thing, for instance https://github.com/conveyal/analysis-ui/blob/dev/lib/components/analysis/spectrogram.js#L24

abyrd replied:
Colors are now parsed with d3-color.

mattwigway commented on lib/dot.js:
```diff
@@ -12,6 +12,9 @@ export default class Dot extends TileLayer.Canvas {
     this.color = new Color(color).rgbArray()
     this.grid = grid
+    //this.colorScale = new CutoffColorScale(1000, [0, 0, 0, 0], [...this.color, 128])
+    this.colorScale = new InterpolatingColorScale([[200, 0, 0, 200, 0], [2000, 0, 0, 200, 180], [20000, 150, 0, 0, 180]])
```
If I understand this correctly, the dot map module now handles all types of maps that require interpolation. We should rename accordingly and add a ColorScale parameter. We should also refactor the existing choropleth code to use the same color scales, unless we decide to drop non-interpolated choropleths altogether.

abyrd replied:
After my last push, it now handles rendering all grids, interpolated or not, with configurable color scales. So yes, it needs to be renamed and have a ColorScale parameter when constructed. I don't fully understand why there's an initialize() method instead of a constructor though.

mattwigway replied:
It's a leaflet layer, and leaflet uses the initialize method, I think because it was built before es6 standardized classes.

abyrd replied:
In what way does leaflet use the initialize method? In my React wrapper I'm instantiating this class myself using a constructor, and it seems to work.

mattwigway replied:
I believe the inherited class constructor calls the initialize method. You might be able to write a constructor with a super call, but this initialize method is how I've always seen leaflet classes written.

See http://leafletjs.com/examples/extending/extending-1-classes.html ,
"In Leaflet’s L.Class, the constructor method is always named initialize."

abyrd replied:
Renaming has been done.

mattwigway commented on lib/dot.js:
```diff
@@ -113,4 +98,176 @@ export default class Dot extends TileLayer.Canvas {
+class CutoffColorScale {
+
+  // any value below cutoff will 
+  // colors are [r, g, b, a]
```
Colors should be parsed with `d3-color`

abyrd replied:
This is done now.

mattwigway reviewed (changes requested):
Nice work, a few comments. Also, going forward please use commitizen format commit messages.

abyrd committed:
switchable bilinear, bicubic, nearest neighbor

can be combined with dots, hard cutoff or interpolated color scales

abyrd commented:
I just saw your comments after pushing some commits. Will switch to commitizen messages for subsequent commits.

abyrd committed:
refactor(interpolation) finish configurable interpolation

Pulled interpolators and color maps out into separate modules.
Renamed classes to reflect the fact that they no longer only make dot maps.

BREAKING CHANGE: This changes the name of the classes that render grids into tiles.
  It also changes the way those classes are configured and initialized.

mattwigway commented on lib/interpolators.js:
```diff
@@ -0,0 +1,173 @@
+// Return the value of an opportunity density grid cell.
+// Range check and return 0 for cells outside the grid.
+// TODO This should really be in the Grid class, and grids should have a default value for empty cells.
+function getGridValue (grid, gx, gy) {
```
In general we should use Javadoc-style comments (`/** . . . */`) to describe classes and methods.

abyrd replied:
Why should we use a Java convention in JS code?

abyrd replied:
Is there some automated docs generation tool?

mattwigway replied:
Yes, JSDoc. We're not currently using it but we might in the future. Also I find that visually differentiating doc comments from comments that describe how the code work is useful for clarity.

abyrd replied:
I have been switching over to this kind of comments.

mattwigway reviewed (changes requested):
Once the `d3-color` changes are in I'd be happy to see this merged. Make sure at least one of the commits has the words BREAKING CHANGE in the body so that a new major version is cut.

mattwigway commented:
Also, it looks like the demo has stopped working with these changes.

trevorgerhardt commented on lib/color-scales.js:
```diff
@@ -0,0 +1,85 @@
+  
+}
+
+export class CutoffColorScale {
```
Using classes for everything is definitely a Java convention :)

The JS-way<sup>TM</sup>:

```js
const cutoffColorScale = (cutoff, lowColor, highColor) =>
  (value) => value < cutoff ? lowColor : highColor
```

This goes for almost all of the classes here assuming you don't need to change the cutoffs or colors on the fly. If you do, might as well just execute again to get the new color function.

abyrd replied:
I just had a conversation with @mattwigway about this too, where he told me that this was our usual convention. Given that JS now has classes though, I am having a hard time understanding why we would want to make second-order functions generating closures instead of simply using named types. It seems like a relic of a more primitive version of the language. Wouldn't classes just transpile to second-order functions anyway?

trevorgerhardt replied:
```js
function cutoff (cutoff, lowColor, highColor) {
  return function (value) {
    return value < cutoff ? lowColor : highColor
  }
}
```

abyrd replied:
@trevorgerhardt you'll be happy to know I've gone full JS-wayTM: The interpolators are now closures (for a 2D patch of pixels) that return more closures for 1D rows of pixels.

trevorgerhardt commented on lib/interpolators.js:
```diff
@@ -0,0 +1,173 @@
+  // Supply an unrolled row-major grid of 16 values (a 4x4 grid).
+  // The resulting object can be used to interpolate between the inner four cells.
+  // Maybe we should be initializing this with a typed 2D array instead of this mess of individual variables.
+  constructor (p00, p01, p02, p03, p10, p11, p12, p13, p20, p21, p22, p23, p30, p31, p32, p33) {
```
For any function with more than two parameters I recommend using options objects.

```js
constructor (grid) {
  this.columnInterpolator0 = new CubicHermiteInterpolator(grid.p00, grid.p01, grid.p02, grid.p03)
  ...
}
```

abyrd replied:
These constructors don't exist anymore, because I'm not using classes.

trevorgerhardt commented:
Don't forget to fix the linting errors!

mattwigway committed:
refactor(classifiers): Rename scales to classifiers to avoid confusion, and add comments.

abyrd committed:
Merge pull request #16 from conveyal/scales-to-classifiers

refactor(classifiers): Rename scales to classifiers to avoid confusio…

abyrd committed:
refactor(classes) convert classes to functions

All inerpolator and color scale classes converted to closures.
Interpolators are now functions that return functions for each row.

abyrd committed:
style(whitespace): remove trailing whitespace and unused imports

abyrd committed:
refactor(BreaksColorScale): remove unused variables, simplify

abyrd committed:
fix(example): reimplement example using only new system

An interpolator, colorizer, and classifier are chosen at will. Classifiers don’t work yet.

BREAKING CHANGE: the existing choropleth layer has been completely removed and several names changed (colorizers, dither, etc.).

abyrd committed:
fix(gitignore): tmp directory was not being ignored

trevorgerhardt commented on lib/colorizers/dither.js:
```diff
@@ -0,0 +1,55 @@
+ * The function returns a single draw from such an approximation centered on the given mean, with the given width.
+ */
+const gaussianApproximation = (iterations, width, mean) => {
+  var sum = 0
```
Generally we should use `let`/`const` in all new code (and prefer `const` when possible).

abyrd replied:
Ah right, I should go through and replace all my vars with lets.

abyrd committed:
feat(Jenks): add a Jenks classifier

This does a much more pleasant-looking job of classifying the data, but is slow.
We need to change the example so it only re-classifies the grid data when that particular property changes.

abyrd committed:
style(Jenks): indentation, whitespace, semicolons

abyrd committed:
style(Jenks): camelCase identifiers, split var decalrations

abyrd committed:
style(whitespace): infix operator spacing, trailing spaces etc.

abyrd committed:
fix(quantile): only skip zeros when they are explicitly passed in as a noDataValue

abyrd committed:
fix(classifiers): getClassIndex is no longer a separate shared function

The test no longer corresponded to any existing function

abyrd committed:
perf(Jenks): apply classifiers only when changed

This is probably not the best way to do this. It would be preferable to do it during / in preparation for rendering, conditionally upon a particular element of state (the classifier) changing.

abyrd committed:
feat(colorizers): Allow supplying D3 colors rather than RGBA

abyrd committed:
refactor(colorizers): express dither as a wrapper on choropleth

This reduces repeated code and more clearly expresses the concept: adding normally distributed noise and then rendering with a choropleth.

abyrd committed:
docs(colorizers): clarify a bit of theory behind the dither colorizer

abyrd committed:
refactor(colorizers): remove unreachable return statements

clamping is now done to the highest break

abyrd committed:
refactor(declarations): replace all var with let

abyrd committed:
feat(colors): allow changing color scheme

abyrd committed:
fix(colors): make original colors same length as others

abyrd committed:
style(comments): changed to c-style comment format

abyrd commented:
All linting errors fixed, and I believe I've taken all review comments into account. The demo works now. Check it out and merge if it looks good!

trevorgerhardt commented on example.js:
```diff
@@ -6,41 +6,133 @@ import React, { Component } from 'react'
 
 export default class GridualizerExample extends Component {
+
+  deStijlColors = [
```
No need to change this now, but we should prefer to define any variables outside of the class when they can be.

abyrd replied:
Meaning anything that's not mutable object instance state should be defined at the top level?

trevorgerhardt replied:
Yup!

trevorgerhardt commented on example.js:
```diff
@@ -6,41 +6,133 @@ import React, { Component } from 'react'
+  }
+
+  // A hand-tweaked hard-wired classifier as a default
+  customClassifier = () => [1500, 10000, 20000, 35000]
```
Placement of definition point applies to these also.

mattwigway commented:
Merged!
